src/macaron/slsa_analyzer/ci_service/github_actions/analyzer.py

```diff
@@ -0,0 +1,718 @@
+    """
+    parsed_obj = job_node.parsed_obj
+    if "Strategy" not in parsed_obj:
+        raise GitHubActionsValueError(f"Unable to find `Strategy` in {job_node.source_path} GitHub Action.")
+    if "Matrix" not in parsed_obj["Strategy"]:
+        raise GitHubActionsValueError(f"Unable to find `Matrix` in {job_node.source_path} GitHub Action.")
+    if "Rows" not in parsed_obj["Strategy"]["Matrix"]:
+        raise GitHubActionsValueError(f"Unable to find `Rows` in {job_node.source_path} GitHub Action.")
+    rows = parsed_obj["Strategy"]["Matrix"]["Rows"]
+    value = rows.get(var)
+    if not value:
+        raise GitHubActionsValueError(f"Unable to find variable {var} in {job_node.source_path} GitHub Action.")
+
+    for item in value["Values"]:
+        yield item["Value"]
+
+
+def is_expression(value: str) -> bool:
+    """Determine if a value is a GitHub Actions expression.
+
+    Parameters
+    ----------
+    value: str
+        The input value.
+
+    Returns
+    -------
+        True if the input value is a GitHub Actions expression.
+
+    Examples
+    --------
+    >>> is_expression("${{ foo }}")
+    True
+    >>> is_expression("${{ foo }")
+    False
+    >>> is_expression("${ foo }")
+    False
+    """
+    return re.match(r"\$\{\{.*?\}\}", value) is not None
+
+
+def find_language_setup_action(job_node: GitHubJobNode, lang_name: BuildLanguage) -> Language | None:
+    """Find the step that calls a language setup GitHub Actions and return the model.
+
+    Parameters
+    ----------
+    job_node: GitHubJobNode
+        The target GitHub Actions job node.
+    lang_name: BuildLanguage
+        The target language used in the build.
+
+    Returns
+    -------
+    Language | None
+        The language model for the language setup GitHub Action or None.
+    """
+    for callee in job_node.callee:
+        model = callee.model
+        # Check if the model implements the Language protocol.
+        if isinstance(model, Language):
+            if model.lang_name == lang_name:
+                return model
+    return None
+
+
+def build_call_graph_from_node(node: GitHubWorkflowNode, repo_path: str) -> None:
+    """Analyze the GitHub Actions node to build the call graph.
+
+    Parameters
+    ----------
+    node : GitHubWorkflowNode
+        The node for a single GitHub Actions workflow.
+    repo_path: str
+        The file system path to the repo.
+    """
+    if not node:
+        return
+
+    for job_name, job in node.parsed_obj.get("Jobs", {}).items():
+        job_node = GitHubJobNode(name=job_name, source_path=node.source_path, parsed_obj=job, caller=node)
+        node.add_callee(job_node)
+        # Add third-party workflows.
+        steps = job.get("Steps") or []
+        for step in steps:
+            if step.get("Exec") and "Uses" in step.get("Exec"):
+                # TODO: change source_path for external workflows.
+                step_name = step["Exec"]["Uses"]["Value"]
```

Comment:
From my understanding `step["Exec"]["Uses"]["Value"]` contains the name of the external workflow.
For example, given this workflow definition:
```yaml
name: test 
on:
  push:
    branches:
      - master
      - '[1-9]+.[0-9]+.x'
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
```

This will be the parsed result for the step `- uses: actions/checkout@v4` (by running the `actionparser` binary against it).

```json
      ...
        {
          "ID": null,
          "Name": null,
          "Exec": {
            "Uses": {
              "Value": "actions/checkout@v4",
              "Quoted": false,
              "Pos": {
                "Line": 11,
                "Col": 15
              }
            },
            "Inputs": null,
            "Entrypoint": null,
            "Args": null
          }
          ...
      }
      ...
```
The name of this step is `null`. Should we rename this variable name to `external_workflow_name` ?

Reply:
Agreed that the `step_name` as variable name here is misleading. I'll change that.

Reply:
See commit [0ac920f](https://github.com/oracle/macaron/pull/673/commits/0ac920f981e277943b32b212350e1b1c65233f41)